File: packages/signing/src/verify.ts
/** @fileoverview Standard Webhooks signature verification. @module @anyhook/signing */
import { createHmac, timingSafeEqual } from 'node:crypto';
import { parseSecret } from './secret.js';

const TOLERANCE_SECONDS = 300;

/** Thrown when a webhook delivery fails signature or timestamp verification. */
export class WebhookVerificationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'WebhookVerificationError';
  }
}

function lowercaseHeaders(headers: Record<string, string>): Record<string, string> {
  const result: Record<string, string> = {};
  for (const [key, value] of Object.entries(headers)) {
    result[key.toLowerCase()] = value;
  }
  return result;
}

function extractSignatureDigest(entry: string): string | undefined {
  const [scheme, digest] = entry.split(',');
  if (scheme !== 'v1' || !digest) {
    return undefined;
  }
  return digest;
}

function timingSafeEqualBase64(a: string, b: string): boolean {
  const bufferA = Buffer.from(a, 'base64');
  const bufferB = Buffer.from(b, 'base64');
  if (bufferA.length !== bufferB.length) {
    return false;
  }
  return timingSafeEqual(bufferA, bufferB);
}

/**
 * Verifies a Standard Webhooks delivery and returns the parsed JSON body.
 * @param headers - The delivery headers (keys are matched case-insensitively)
 * @param rawBody - The exact raw request body string
 * @param secret - The signing secret (with or without `whsec_` prefix)
 * @returns The parsed JSON body
 * @throws {WebhookVerificationError} If headers are missing, the timestamp is
 *   outside tolerance, or no signature matches
 */
export function verify(headers: Record<string, string>, rawBody: string, secret: string): unknown {
  const lower = lowercaseHeaders(headers);
  const id = lower['webhook-id'];
  const timestamp = lower['webhook-timestamp'];
  const signature = lower['webhook-signature'];

  if (!id || !timestamp || !signature) {
    throw new WebhookVerificationError('Missing required webhook headers');
  }

  const timestampSeconds = Number(timestamp);
  if (!Number.isFinite(timestampSeconds)) {
    throw new WebhookVerificationError('Invalid webhook timestamp');
  }

  const nowSeconds = Math.floor(Date.now() / 1000);
  if (Math.abs(nowSeconds - timestampSeconds) > TOLERANCE_SECONDS) {
    throw new WebhookVerificationError('Webhook timestamp is outside of tolerance');
  }

  const signedContent = `${id}.${timestampSeconds}.${rawBody}`;
  const key = parseSecret(secret);
  const expectedDigest = createHmac('sha256', key).update(signedContent).digest('base64');

  const providedSignatures = signature.split(' ');
  for (const entry of providedSignatures) {
    const digest = extractSignatureDigest(entry);
    if (digest && timingSafeEqualBase64(digest, expectedDigest)) {
      return JSON.parse(rawBody);
    }
  }

  throw new WebhookVerificationError('No matching signature found');
}

File: go/signing/verify.go
package signing

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"fmt"
	"math"
	"strconv"
	"strings"
	"time"
)

// ToleranceSeconds is the allowed clock-skew window for Verify.
const ToleranceSeconds = 300

// WebhookVerificationError is returned when a webhook delivery fails
// signature or timestamp verification.
type WebhookVerificationError struct {
	msg string
}

func (e *WebhookVerificationError) Error() string { return e.msg }

func verificationErr(msg string) error { return &WebhookVerificationError{msg: msg} }

func lowercaseHeaders(headers map[string]string) map[string]string {
	out := make(map[string]string, len(headers))
	for k, v := range headers {
		out[strings.ToLower(k)] = v
	}
	return out
}

func extractSignatureDigest(entry string) (string, bool) {
	scheme, digest, found := strings.Cut(entry, ",")
	if !found || scheme != "v1" || digest == "" {
		return "", false
	}
	return digest, true
}

func timingSafeEqualBase64(a, b string) bool {
	bufA, errA := base64.StdEncoding.DecodeString(a)
	bufB, errB := base64.StdEncoding.DecodeString(b)
	if errA != nil || errB != nil {
		return false
	}
	if len(bufA) != len(bufB) {
		return false
	}
	return hmac.Equal(bufA, bufB)
}

// Verify verifies a Standard Webhooks delivery and returns the raw, verified
// body bytes.
//
//   - headers: the delivery headers (keys matched case-insensitively)
//   - rawBody: the exact raw request body string
//   - secret: the signing secret (with or without the whsec_ prefix)
//
// Returns a *WebhookVerificationError if headers are missing, the timestamp
// is outside tolerance, or no signature matches.
//
// Unlike the TS verify() (which returns the JSON.parse()'d body as `unknown`),
// Verify returns the raw validated bytes -- more idiomatic Go, letting the
// caller json.Unmarshal into whatever type it expects.
func Verify(headers map[string]string, rawBody string, secret string) ([]byte, error) {
	lower := lowercaseHeaders(headers)
	id := lower["webhook-id"]
	timestamp := lower["webhook-timestamp"]
	signature := lower["webhook-signature"]

	if id == "" || timestamp == "" || signature == "" {
		return nil, verificationErr("Missing required webhook headers")
	}

	timestampSeconds, err := strconv.ParseFloat(timestamp, 64)
	if err != nil || math.IsNaN(timestampSeconds) || math.IsInf(timestampSeconds, 0) {
		return nil, verificationErr("Invalid webhook timestamp")
	}

	nowSeconds := float64(time.Now().Unix())
	if math.Abs(nowSeconds-timestampSeconds) > ToleranceSeconds {
		return nil, verificationErr("Webhook timestamp is outside of tolerance")
	}

	signedContent := fmt.Sprintf("%s.%d.%s", id, int64(timestampSeconds), rawBody)
	key, err := ParseSecret(secret)
	if err != nil {
		return nil, verificationErr("Invalid signing secret")
	}
	mac := hmac.New(sha256.New, key)
	mac.Write([]byte(signedContent))
	expectedDigest := base64.StdEncoding.EncodeToString(mac.Sum(nil))

	for _, entry := range strings.Split(signature, " ") {
		digest, ok := extractSignatureDigest(entry)
		if ok && timingSafeEqualBase64(digest, expectedDigest) {
			return []byte(rawBody), nil
		}
	}

	return nil, verificationErr("No matching signature found")
}
